File: template/.claude/hooks/pre_edit.py
#!/usr/bin/env python3
"""Pre-Edit hook: protects files and enforces LOC limits.

Hard blocks:
- Files outside Allowed Files in spec (when spec exists)
- Protected test files (contracts/, regression/)

Soft blocks:
- Files exceeding LOC limits (400 code, 600 tests)
"""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from utils import (
    allow_tool,
    ask_tool,
    deny_tool,
    get_tool_input,
    infer_spec_from_branch,
    is_file_allowed,
    read_hook_input,
)

# Protected paths (Hard Block)
PROTECTED_PATHS = [
    "tests/contracts/",
    "tests/regression/",
]

# LOC limits (Soft Block)
MAX_LOC_CODE = 400
MAX_LOC_TEST = 600
# 7/8 = 0.875 -> gives round thresholds: 350 LOC (code), 525 LOC (tests)
WARN_THRESHOLD = 0.875

# Sync zones (files that should stay in sync with template/)
SYNC_ZONES = [
    ".claude/",
    "scripts/",
]

# Excluded from sync reminders (DLD-specific customizations)
EXCLUDE_FROM_SYNC = [
    ".claude/rules/localization.md",
    ".claude/rules/template-sync.md",
    ".claude/rules/git-local-folders.md",
    ".claude/CUSTOMIZATIONS.md",
    ".claude/settings.local.json",
]


def count_lines(file_path: str) -> int:
    """Count lines in file."""
    try:
        with open(file_path, encoding="utf-8", errors="ignore") as f:
            return sum(1 for _ in f)
    except Exception:
        return 0


def is_test_file(file_path: str) -> bool:
    """Check if file is a test file."""
    return "_test.py" in file_path or "/tests/" in file_path


def normalize_path(file_path: str) -> str:
    """Convert absolute path to relative."""
    if not file_path:
        return ""

    project_dir = os.environ.get("CLAUDE_PROJECT_DIR", os.getcwd())
    if file_path.startswith(project_dir):
        return file_path[len(project_dir) :].lstrip("/")
    return file_path


def check_sync_zone(rel_path: str) -> "str | None":
    """Returns reminder message if file is in sync zone."""
    if not rel_path:
        return None

    # Check if file is in a sync zone
    in_sync_zone = False
    for zone in SYNC_ZONES:
        if rel_path.startswith(zone):
            in_sync_zone = True
            break

    if not in_sync_zone:
        return None

    # Check if file is excluded from sync
    if rel_path in EXCLUDE_FROM_SYNC:
        return None

    # Check if template version exists
    template_path = f"template/{rel_path}"
    if os.path.exists(template_path):
        return (
            f"SYNC ZONE: {rel_path}\n\n"
            f"This file exists in template/{rel_path}\n"
            f"Remember to sync changes bidirectionally.\n\n"
            f"See: .claude/rules/template-sync.md"
        )

    return None


def _log_error(error: Exception) -> None:
    """Log hook error for diagnostics."""
    try:
        import datetime

        with open("/tmp/claude-hook-errors.log", "a") as f:  # nosec B108
            f.write(f"{datetime.datetime.now()} [pre_edit]: {error}\n")
    except Exception:
        pass  # nosec B110 - intentional fail-safe


def main():
    try:
        data = read_hook_input()
        file_path = get_tool_input(data, "file_path") or ""

        rel_path = normalize_path(file_path)

        # Check Allowed Files (Hard Block) - only when spec exists
        spec_path = os.environ.get("CLAUDE_CURRENT_SPEC_PATH") or infer_spec_from_branch()
        allowed, allowed_files = is_file_allowed(rel_path, spec_path)
        if not allowed:
            allowed_list = "\n".join(f"  - {f}" for f in allowed_files[:10])
            deny_tool(
                f"File not in Allowed Files!\n\n"
                f"{rel_path}\n\n"
                f"Spec: {spec_path or '(not found)'}\n\n"
                f"Allowed files:\n{allowed_list}\n\n"
                f"Add file to spec's ## Allowed Files or change approach."
            )
            return

        # Check protected paths (Hard Block)
        for protected in PROTECTED_PATHS:
            if protected in rel_path:
                deny_tool(
                    f"Protected test file!\n\n"
                    f"{rel_path}\n\n"
                    f"tests/contracts/ and tests/regression/ cannot be modified.\n"
                    f"Fix the code, not the test.\n\n"
                    f"See: CLAUDE.md -> Test Safety"
                )
                return

        # Check LOC limits (Soft Block)
        # Use absolute path for file operations
        abs_path = file_path if file_path.startswith("/") else os.path.join(os.getcwd(), file_path)

        if os.path.exists(abs_path):
            loc = count_lines(abs_path)
            max_loc = MAX_LOC_TEST if is_test_file(rel_path) else MAX_LOC_CODE
            warn_loc = int(max_loc * WARN_THRESHOLD)

            if loc >= max_loc:
                ask_tool(
                    f"File exceeds LOC limit!\n\n"
                    f"{rel_path}: {loc} lines (limit: {max_loc})\n\n"
                    f"Consider splitting the file.\n"
                    f"See: CLAUDE.md -> File Limits\n\n"
                    f"Proceed anyway?"
                )
                return
            elif loc >= warn_loc:
                ask_tool(
                    f"File approaching LOC limit\n\n"
                    f"{rel_path}: {loc} lines (limit: {max_loc})\n\n"
                    f"Proceed?"
                )
                return

        # Check sync zone (Soft reminder - just info, doesn't block)
        sync_reminder = check_sync_zone(rel_path)
        if sync_reminder:
            # Use ask_tool to show reminder but allow proceeding
            ask_tool(sync_reminder)
            return

        allow_tool()
    except Exception as e:
        _log_error(e)
        allow_tool()


if __name__ == "__main__":
    main()

File: template/.claude/hooks/pre-edit.mjs
/**
 * Pre-Edit hook: protects files and enforces LOC limits.
 *
 * Hard blocks:
 * - Files outside Allowed Files in spec (when spec exists)
 * - Protected test files (contracts/, regression/)
 * - Plan-before-code gate (src/ edits blocked when autopilot-state.json has plan_exists: false)
 *
 * Soft blocks:
 * - Files exceeding LOC limits (400 code, 600 tests)
 *
 * Configurable via hooks.config.mjs / hooks.config.local.mjs
 */

import { readFileSync, existsSync } from 'fs';
import { join } from 'path';
import {
  allowTool,
  askTool,
  debugLog,
  debugTiming,
  denyTool,
  getProjectDir,
  getToolInput,
  getWorktreeRoot,
  inferSpecFromBranch,
  isFileAllowed,
  loadConfig,
  logHookError,
  readHookInput,
} from './utils.mjs';

// Hardcoded fallbacks (used when config is unavailable)
const FALLBACK_PROTECTED = ['tests/contracts/', 'tests/regression/'];
const FALLBACK_MAX_LOC_CODE = 400;
const FALLBACK_MAX_LOC_TEST = 600;
const FALLBACK_WARN_THRESHOLD = 7 / 8;
const FALLBACK_SYNC_ZONES = ['.claude/', 'scripts/'];
const FALLBACK_EXCLUDE_SYNC = [
  '.claude/rules/localization.md',
  '.claude/CUSTOMIZATIONS.md',
  '.claude/settings.local.json',
];
const FALLBACK_INT_PATTERNS = [/^tests\/integration\//, /\.integration\.test\./, /\.integration\.spec\./];
const FALLBACK_MOCK_PATTERNS = [/jest\.mock\s*\(/, /vi\.mock\s*\(/, /\bunittest\.mock\b/, /\bMagicMock\b/, /@patch\b/, /\bmock\.patch\b/, /\bsinon\.stub\b/, /\bsinon\.mock\b/];

function countLines(filePath) {
  try {
    const content = readFileSync(filePath, 'utf-8');
    if (!content) return 0;
    return content.split('\n').length - (content.endsWith('\n') ? 1 : 0);
  } catch {
    return 0;
  }
}

/**
 * Whether the pending edit leaves the file no larger than it already is.
 *
 * A file already past the limit is otherwise unfixable: every edit is denied,
 * while the hook's own advice ("split the file") requires editing that same
 * file. Growth stays blocked; shrinking edits — the split itself, deletions,
 * refactors — are allowed, so the advice becomes possible to follow.
 *
 * For Edit, comparing the line counts of old_string and new_string is enough,
 * and it holds for replace_all too: a delta that is non-positive per occurrence
 * cannot grow the file however many occurrences are replaced.
 *
 * Note: getToolInput() returns null for an empty string, so a deletion arrives
 * as old_string set and new_string null — that is a shrink, not a missing key.
 */
function editDoesNotGrowFile(data, currentLoc) {
  const oldString = getToolInput(data, 'old_string');
  if (oldString !== null) {
    const newString = getToolInput(data, 'new_string') || '';
    const removed = oldString.split('\n').length;
    const added = newString ? newString.split('\n').length : 0;
    return added <= removed;
  }

  const content = getToolInput(data, 'content');
  if (content !== null) {
    const newLoc = content.split('\n').length - (content.endsWith('\n') ? 1 : 0);
    return newLoc <= currentLoc;
  }

  return false;
}

const TEST_FILE_PATTERNS = [
  /_test\./, /\.test\./, /\.spec\./,
  /\/tests?\//, /__tests__\//,
  /(^|\/)test_[^/]+\.py$/,
];

function isTestFile(filePath) {
  return TEST_FILE_PATTERNS.some(pattern => pattern.test(filePath));
}

function isIntegrationTest(relPath, patterns) {
  return patterns.some(p => p.test(relPath));
}

function containsMockPattern(content, patterns) {
  if (!content) return null;
  for (const p of patterns) {
    if (p.test(content)) return p.source;
  }
  return null;
}

/**
 * Find autopilot-state.json in project dir or worktree root.
 * Covers mismatch when CLAUDE_PROJECT_DIR != cwd worktree.
 */
function findAutopilotState() {
  const candidates = [join(getProjectDir(), 'autopilot-state.json')];
  const worktreeRoot = getWorktreeRoot();
  if (worktreeRoot) {
    candidates.push(join(worktreeRoot, 'autopilot-state.json'));
  }
  for (const p of candidates) {
    if (existsSync(p)) {
      try { return JSON.parse(readFileSync(p, 'utf-8')); } catch { /* fail-safe */ }
    }
  }
  return null;
}

function normalizePath(filePath) {
  if (!filePath) return '';
  const projectDir = getProjectDir();
  if (filePath.startsWith(projectDir)) {
    return filePath.slice(projectDir.length).replace(/^[/\\]/, '');
  }
  return filePath;
}

function checkSyncZone(relPath, syncZones, excludeFromSync) {
  if (!relPath) return null;

  const inSyncZone = syncZones.some(zone => relPath.startsWith(zone));
  if (!inSyncZone) return null;
  if (excludeFromSync.includes(relPath)) return null;

  const templatePath = join(getProjectDir(), 'template', relPath);
  if (existsSync(templatePath)) {
    return (
      `SYNC ZONE: ${relPath}\n\n` +
      `This file exists in template/${relPath}\n` +
      `Remember to sync changes bidirectionally.\n\n` +
      `See: .claude/rules/template-sync.md`
    );
  }
  return null;
}

async function main() {
  const timer = debugTiming('pre-edit');
  try {
    const data = readHookInput();
    const filePath = getToolInput(data, 'file_path') || '';
    const relPath = normalizePath(filePath);
    debugLog('pre-edit', 'input', { file: relPath });

    const config = await loadConfig();
    const protectedPaths = config?.preEdit?.protectedPaths || FALLBACK_PROTECTED;
    const maxLocCode = config?.preEdit?.maxLocCode ?? FALLBACK_MAX_LOC_CODE;
    const maxLocTest = config?.preEdit?.maxLocTest ?? FALLBACK_MAX_LOC_TEST;
    const warnThreshold = config?.preEdit?.warnThreshold ?? FALLBACK_WARN_THRESHOLD;
    const syncZones = config?.preEdit?.syncZones || FALLBACK_SYNC_ZONES;
    const excludeFromSync = config?.preEdit?.excludeFromSync || FALLBACK_EXCLUDE_SYNC;
    const alwaysAllowed = config?.utils?.alwaysAllowedPatterns || undefined;

    // Check Allowed Files (Hard Block) - only when spec exists
    const specPath = process.env.CLAUDE_CURRENT_SPEC_PATH || inferSpecFromBranch();
    const { allowed, allowedFiles } = isFileAllowed(relPath, specPath, alwaysAllowed);
    if (!allowed) {
      const allowedList = allowedFiles.slice(0, 10).map(f => `  - ${f}`).join('\n');
      debugLog('pre-edit', 'deny', { reason: 'not_in_allowed_files', file: relPath });
      timer.end('deny');
      denyTool(
        `File not in Allowed Files!\n\n` +
          `${relPath}\n\n` +
          `Spec: ${specPath || '(not found)'}\n\n` +
          `Allowed files:\n${allowedList}\n\n` +
          `To fix:\n` +
          `1. Edit ${specPath || '(spec file)'}\n` +
          `2. Find ## Allowed Files section\n` +
          `3. Add: \`${relPath}\` — {description}\n` +
          `4. Save and retry\n\n` +
          `Or change approach to use only allowed files.`,
      );
      return;
    }

    // Check protected paths (Hard Block)
    for (const protectedPath of protectedPaths) {
      if (relPath.startsWith(protectedPath)) {
        debugLog('pre-edit', 'deny', { reason: 'protected_path', file: relPath });
        timer.end('deny');
        denyTool(
          `Protected test file!\n\n` +
            `${relPath}\n\n` +
            `tests/contracts/ and tests/regression/ cannot be modified.\n` +
            `Fix the code, not the test.\n\n` +
            `See: CLAUDE.md -> Test Safety`,
        );
        return;
      }
    }

    // Check mock ban in integration tests (Hard Block)
    const mockBan = config?.preEdit?.mockBan;
    if (mockBan?.enabled !== false) {
      const intPatterns = mockBan?.integrationTestPatterns || FALLBACK_INT_PATTERNS;
      if (isIntegrationTest(relPath, intPatterns)) {
        const newContent = getToolInput(data, 'new_string') || getToolInput(data, 'content') || '';
        const mockPats = mockBan?.mockPatterns || FALLBACK_MOCK_PATTERNS;
        const matched = containsMockPattern(newContent, mockPats);
        if (matched) {
          debugLog('pre-edit', 'deny', { reason: 'mock_in_integration', file: relPath, pattern: matched });
          timer.end('deny');
          denyTool(
            `Mock in integration test!\n\n` +
              `${relPath}\nPattern: ${matched}\n\n` +
              `Integration tests must use real dependencies (Testcontainers).\n` +
              `Mocks allowed only in tests/unit/.`,
          );
          return;
        }
      }
    }

    // Check plan-before-code gate (Hard Block)
    // Only activates when autopilot-state.json exists AND plan_exists is false
    const requirePlan = config?.enforcement?.requirePlanBeforeCode !== false;
    if (requirePlan && relPath.startsWith('src/')) {
      try {
        const autopilotState = findAutopilotState();
        if (autopilotState && autopilotState.plan_exists === false) {
          debugLog('pre-edit', 'deny', { reason: 'no_plan', file: relPath });
          timer.end('deny');
          denyTool(
            'Plan not found in spec. Run planner first.\n\n' +
              `File: ${relPath}\n` +
              'autopilot-state.json shows plan_exists: false\n\n' +
              'The planner must create an implementation plan before code changes.\n' +
              'See: task-loop.md -> Step 1',
          );
          return;
        }
      } catch {
        // fail-safe: autopilot-state read error = don't block (ADR-004)
      }
    }

    // Check LOC limits (Soft Block)
    const absPath = filePath.startsWith('/') ? filePath : join(process.cwd(), filePath);

    if (existsSync(absPath)) {
      const loc = countLines(absPath);
      const maxLoc = isTestFile(relPath) ? maxLocTest : maxLocCode;
      const warnLoc = Math.floor(maxLoc * warnThreshold);

      if (loc >= maxLoc) {
        if (editDoesNotGrowFile(data, loc)) {
          debugLog('pre-edit', 'allow', { reason: 'loc_limit_shrinking_edit', file: relPath, loc, maxLoc });
          timer.end('allow');
          allowTool();
          return;
        }
        debugLog('pre-edit', 'deny', { reason: 'loc_limit', file: relPath, loc, maxLoc });
        timer.end('deny');
        denyTool(
          `File exceeds LOC limit!\n\n` +
            `${relPath}: ${loc} lines (limit: ${maxLoc})\n\n` +
            `Edits that grow this file are blocked. Edits that shrink it are allowed,\n` +
            `so the split itself is possible: move code into a new module, then delete\n` +
            `it here. Deletions and same-size rewrites pass too.\n\n` +
            `See: CLAUDE.md -> File Limits`,
        );
        return;
      } else if (loc >= warnLoc) {
        // Soft warning: allow edit but log — denyTool() would be too aggressive,
        // askTool() kills bypass mode (claude-code#37420)
        debugLog('pre-edit', 'allow', { reason: 'loc_warning', file: relPath, loc, maxLoc });
        timer.end('allow');
        allowTool();
        return;
      }
    }

    // Check sync zone (Soft reminder) — allow edit, log reminder
    // askTool() kills bypass mode (claude-code#37420), and denyTool() would block all template edits
    const syncReminder = checkSyncZone(relPath, syncZones, excludeFromSync);
    if (syncReminder) {
      debugLog('pre-edit', 'allow', { reason: 'sync_zone', file: relPath, reminder: syncReminder });
      timer.end('allow');
      allowTool();
      return;
    }

    debugLog('pre-edit', 'allow', { file: relPath });
    timer.end('allow');
    allowTool();
  } catch (e) {
    debugLog('pre-edit', 'error', { error: String(e) });
    timer.end('error');
    logHookError('pre_edit', e);
    allowTool();
  }
}

main();
